gym_sapientino/core/grid.py

```diff
@@ -0,0 +1,101 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019-2020 Marco Favorito, Luca Iocchi
+#
+# ------------------------------
+#
+# This file is part of gym-sapientino.
+#
+# gym-sapientino is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# gym-sapientino is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with gym-sapientino.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+"""Classes to represent a Sapientino map."""
+from pathlib import Path
+from typing import Dict, Iterator, List
+
+from gym_sapientino.core.types import Colors, color2int, id2color
+
+
+class Cell:
+    """A class to represent a cell on the grid."""
+
+    def __init__(self, x: int, y: int, color: Colors):
+        """Initialize the cell."""
+        self.x = x
+        self.y = y
+        self.color = color
+        self.bip_count = 0
+
+    @property
+    def encoded_color(self) -> int:
+        """Encode the color."""
+        return color2int[self.color]
+
+    def beep(self) -> None:
+        """Do a beep."""
+        self.bip_count += 1
+
+
+class SapientinoGrid:
+    """The grid of the Sapientino environment."""
+
+    def __init__(self, cells: List[List[Cell]]):
+        """Initialize the grid."""
+        self.cells: List[List[Cell]] = cells
+        self.color_count: Dict[Colors, int] = {}
+
+    @property
+    def rows(self):
+        """Get the number of rows."""
+        # TODO allow different number of rows and columns.
+        return len(self.cells)
```

Comment:
What do you mean by "different number of rows and columns"?
Non rectangular map? That yould be also nice, and not too hard either, but I don't know if the transition functions assume a square shape somewhere.

Reply:
no, Sapientino doesn't assume a square shape. You're right, it's not that hard, but maybe for our purposes, this is good enough. We will see

Reply:
It is! we don't need it now. So rows and columns are correct, right? Why the TODOs.
Also, in the parsing function, you could do something like `all((len(cell_str[i]) == nb_columns for i in nb_rows))`
Nothing important btw

Reply:
Will fix in `from_map` during the inner for loop